### PR TITLE
Logging interface

### PR DESCRIPTION
**Pull request by clalancette**
As discussed in [several](https://github.com/ros2/rcl_logging/pull/33#issuecomment-629416344) [places](https://github.com/ros2/rcl_logging/pull/40#discussion_r435265222), the current way that the rcl_logging API is working is quite odd.  The API is currently being defined in rcl, implemented in rcl_logging, and then used in rcl.  This essentially creates a circular dependency that we are breaking by copying a bunch of code around.

Instead of that, what this PR (and the counterpart in ros2/rcl#676) does is to define an `rcl_logging_interface` package.  This package *only* installs and provides a header file.  Then the logging backends depend on this interface and implement it.  Finally, rcl also depends on the interface, and calls into whatever backend is setup as the default.  This breaks the circular dependency.

@rotu @wjwwood @cottsay @hidmic FYI.

### CODE REVIEW

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11047/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6371/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8998/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10970/)

File: rcl_logging_interface/include/rcl_logging_interface/visibility_control.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCL_LOGGING_INTERFACE__VISIBILITY_CONTROL_H_
+#define RCL_LOGGING_INTERFACE__VISIBILITY_CONTROL_H_
+
+#ifdef __cplusplus
+extern "C"
+{
```
Comment (hidmic) [inline rcl_logging_interface/include/rcl_logging_interface/visibility_control.h:20]: @clalancette nit^2: be consistent with the curly bracket positioning. Some files keep it at the end of the line where `extern "C"` shows, others do this.

Comment (clalancette) [reply rcl_logging_interface/include/rcl_logging_interface/visibility_control.h]: 04ebe64bc75d3f3dc049e9e9e4b86b3f4f58c5bf should improve the situation.

File: rcl_logging_noop/CMakeLists.txt [hunk 2]
```diff
@@ -46,6 +43,6 @@ install(TARGETS ${PROJECT_NAME}
   RUNTIME DESTINATION bin)
```
Comment (cottsay) [inline rcl_logging_noop/CMakeLists.txt:43]: Does this `install()` need `EXPORT ${PROJECT_NAME}` like the other two logging implementations?

Comment (clalancette) [reply rcl_logging_noop/CMakeLists.txt:43]: Oh yeah, probably.  Also probably needs `ament_export_target`.  Will add both.

**Comment by clalancette** [comment]
> Thanks, Chris, this is great. Do we need to add a quality declaration since this new package will be in the "critical path"?

Yeah, good point.  Added in 655ee72.

**Comment by clalancette** [comment]
All right, smoke-test CI looked good.  Here is a full CI to make sure nothing wonky happens:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11069/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6382/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9014/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10983/)

**Comment by clalancette** [comment]
All right, Linux warnings are expected and in the nightlies.

macOS failures are all in the nightlies except one.  But given that it is a failure in a service, I think that is probably flaky.

Windows is hard to tell, but there are 4 additional tests that failed here vs. the nightlies.  All of those were in services as well, so I think we can chalk that up to flakiness.

Overall, it looks like this set of PRs doesn't have much impact (which is good!).  Going ahead and merging, thanks for the reviews.

**Pull request merged**